src/ScriptEditor/ScriptEditor.ts

```diff
@@ -63,26 +63,45 @@ export class ScriptEditor {
 
     // Add ts definitions for API
     const source = netscriptDefinitions.replace(/export /g, "");
-    for (const languageDefaults of [
-      monaco.languages.typescript.javascriptDefaults,
-      monaco.languages.typescript.typescriptDefaults,
-    ]) {
+    for (const [language, languageDefaults] of [
+      ["javascript", monaco.languages.typescript.javascriptDefaults],
+      ["typescript", monaco.languages.typescript.typescriptDefaults],
+    ] as const) {
       languageDefaults.addExtraLib(source, "netscript.d.ts");
       languageDefaults.addExtraLib(reactTypes, "react.d.ts");
       languageDefaults.addExtraLib(reactDomTypes, "react-dom.d.ts");
+      languageDefaults.setCompilerOptions({
+        ...languageDefaults.getCompilerOptions(),
+        // We allow direct importing of `.ts`/`.tsx` files, so tell the typescript language server that.
+        allowImportingTsExtensions: true,
+        // We use file-at-a-time transpiler. See https://www.typescriptlang.org/tsconfig/#isolatedModules
+        isolatedModules: true,
+        // We use the classic (i.e. `React.createElement`:) react runtime.
+        jsx: monaco.languages.typescript.JsxEmit.React,
+        // We define `React` and `ReactDOM` as globals. Don't mark using them as errors.
+        allowUmdGlobalAccess: true,
+        // Enable strict typechecking.
+        // Note that checking in javascript is disabled by default but can be enabled via `// @ts-check`.
+        // This enables strictNullChecks, which impacts reported types, even in javascript.
+        strict: true,
+        noImplicitAny: language === "typescript",
```

Comment:
I think this is the only unresolved point that catlover brought up. Why disable this specifically for JS?

Reply:
I don't have an entirely coherent reason for this, but there are a number of things that lead to this choice (though I'm not sure I had considered all this explicitly when I made the choice):
- I was considering the possibility of having an option to turn `checkJS` on globally (likely in the editor options), and though that this would be a better default for that case.
- Given that javascript code is also (I think anyway), syntactically valid typescript code, and we support calling freely between javascript and typescript files, there is an easy way for people get the stricter typechecking, and having different settings for javascript and typescript allows people to choose which strictness they want.
- Partly I thought it would be nice for how I'd be inclined to migrate my code, once this gets to a release:
  - First turn on checking for types that it can deduce or are explicitly delcared, but letting code that doesn't have  types stay as-is.
  - Second, go through and add types for things that aren't already deduced.

Reply:
I went through and re-read more carefully what `// @ts-check` does. Some notes:
* Unless I got overruled by Snarling, I would reject a global `checkJS` option. Anything that alters the behavior of *code* should be configured in code (or in a file-based config), so that it can be shared if need be. I've encountered too many issues in other games/environments where options and environment settings cause code that works for one person to not work for others; it's the devil, as far as I'm concerned.
* For migration, I think the best thing to do would be to work file-at-a-time and just rename your .js file to .ts, then fixup any type errors that occur. You can bail out to `any` if you need a quick escape hatch for a big section.

That said, `// @ts-check` is a sort of halfway point that will never do a great job of typechecking, because the JSDoc syntax it relies on just isn't expressive enough for that. So I'm OK with leaving the default as implicit-any.